### ISSUE TITLE
Provide a low CPU usage mode for laptops and to avoid annoying fan spin ups all the time during idle on desktop

### ISSUE THREAD

**Issue by JonasT** [syncthing/syncthing#4864]
syncthing appears to reduce its own priority with nice, which means I've never had trouble with it making *other* programs run slow while it runs in the background - which is very good!

However, it will still max out one core to 100% use and make use of other cores approximately twice a minute, even when *no other device is connected at all* and there are no folder changes - I assume for folder scanning.

Why is this a problem? On my laptop, this causes an audible fan spin up from nothing to a notable volume. The laptop is idle, has nothing to do, I'm not doing anything with it myself, no files are changed - and still the fan constantly spins up, making noise, wasting energy and most likely reducing the runtime on battery notably for no good reason (after all, there are no changes and there is nothing to sync!). It is even an issue on my desktop, because again during idle, there are notable fan speed changes which can simply be a psychological issue - it's quite annoying to listen to.

Therefore, I suggest that if there are no devices connected and there are are no notable folder changes for a longer time, syncthing should limit its own CPU usage to prevent such issues. (with cgroup or something maybe?) It could be an option so that power users who want to keep the maximum folder scanning speed could disable it if they want to.

### Version Information

Syncthing Version:  v0.14.46, Linux (64 bit) 
OS Version: Fedora 27 x64

**Comment by AudriusButkevicius** [comment]
I am closing this as this should be first talked about in the forum. Syncthing shoyld not use CPU if nothing has changed. If it is using CPU it means it is scanning, and if it is scanning, then something is changing.

**Comment by calmh** [comment]
Just doing the walk will use some amount of cpu and io even if nothing has changed; using the file system notifications is the solution to this.

**Comment by JonasT** [comment]
Yea, it's definitely maxing out the CPU @AudriusButkevicius .

Are file system notifications an upcoming feature? I think I saw them in the options somewhere - if they are the solutions to this, is there a specific reason they aren't used per default?

If I remember vaguely, Linux file notifications (inotify) don't work too well with huge folders. Maybe it would be still worth solving this without them, e.g. with cgroup CPU limit or something when not much is happening? I still think that's one of the most straightforward ways to avoid this problem.

**Comment by AudriusButkevicius** [comment]
@JonasT as I said, all of this should be discussed on the forum.
Github issue tracker is not the right place for generic questions.

**Comment by JonasT** [comment]
Is this the wrong place for feature requests? Sorry about that, most projects want those on the issue tracker.

**Comment by AudriusButkevicius** [comment]
This is the right place for feature requests, but only after they have been discussed, fleshed out and agreed with, and all of the latter happens on the forum.

**Comment by JonasT** [comment]
Just for the record, my feature request is apply a CPU limit with cgroup (or similar) when there is no sync activity, to avoid CPU spikes outside of that (e.g. during scanning). If you want to reject it for now, that's fine - just wanted to be sure it's obvious what I'm suggesting.

**Comment by calmh** [comment]
I think that’s valid to discuss here. However I’m not sure it belongs in Syncthing, as it’s a policy enforced from the outside. We provide two controls for CPU usage already, in the form of GOMAXPROCS and the built in nicing. In the normal case we want things to happen as quickly as possible, so throttling cpu usage isn’t useful. The OS can of course impose other limits based on user configuration.

**Comment by JonasT** [comment]
Sure but neither of them will prevent a core from getting fully exhausted. (as I mentioned above, `nice` just makes sure syncthing is nice to other programs, it will never limit the *absolute* CPU usage as far as I'm aware.) OS isn't really useful here either (IMHO) because it's too complicated for the average user. What would be nice would be a simple "limit CPU usage during idle with no sync activity" option that did this automatically.

**Comment by dcarosone** [comment]
When there are no peers connected for a folder (or at all) , we could put
that folder (or the daemon) into an idled state.

When a folder is idled, is there value in doing scans at all?

Can they be done at a slower interval (say 3x, or some backoff scheme if no
local changes are seen on each pass)?

Can / should connection of a peer trigger a scan in folders shared with
that peer, if it was previously idled? Maybe before exchanging state?

Using watchers is great, I've been using them with syncthing-gtk for ages,
but not a complete solution; I seem to hit kernel limits on the number of
watches and have to be selective.

**Comment by Ferroin** [comment]
@JonasT There really isn't a good way to do it from inside Syncthing using cgroups either, at least not protably across Linux distributions.  Control groups are as a general rule managed by administrative processes (and if you're using systemd, they're managed by systemd, _period_, otherwise you run into all kinds of problems), and aren't designed for management by more than one process (unless you're using delegation, but that's complicated, error-prone, and isn't widely supported).

Assuming you're on a system which doesn't use systemd, you could in theory just write up a simple script to poll the appropriate REST API endpoint and check if anything is connected, and if not move syncthing to a lower priority cgroup.  The hardest part there is handling the REST API and finding the right balance between responsiveness (so your system isn't sluggish when you actually need things to sync) and resource efficiency.

**Comment by JonasT** [comment]
> Can they be done at a slower interval

That doesn't help in some cases, because when the folder is big enough the single scan itself will be enough to cause considerable CPU load. It might fix the battery issue more or less, but not the fan spin up / general unevenness of load. syncthing is a unique position here, I don't have any other system service that exhibits behavior like this (they either cause big load very rarely like the update services, or usually not much of it - not regularly high load like syncthing every minute), so that's why I think it might be worth putting something into it to mitigate that.

> There really isn't a good way to do it from inside Syncthing using cgroups either, at least not protably across Linux distributions. Control groups are as a general rule managed by administrative processes

True. I did just try `cpulimit` on syncthing however, and it works *amazingly*. I'm not sure how exactly it works, but it seems to be some sort of Unix signaling stuff. Maybe that would be a better way?
